File: app-tauri/src-tauri/src/address_attempts.rs
//! Pure deterministic address-attempt planning with no DNS or socket I/O.

use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};

#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum AddressFamily {
    Ipv4,
    Ipv6,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct AddressCandidate {
    pub id: String,
    pub family: AddressFamily,
    pub address: String,
    pub source: String,
    pub expires_at_ms: Option<u64>,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum AddressAttemptState {
    Running,
    Failed,
    Succeeded,
    Cancelled,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct AddressAttempt {
    pub candidate_id: String,
    pub state: AddressAttemptState,
    pub started_at_ms: u64,
    pub completed_at_ms: Option<u64>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct AddressPlanContext {
    pub now_ms: u64,
    pub started_at_ms: u64,
    pub deadline_at_ms: u64,
    pub stagger_ms: u64,
    pub max_concurrent: usize,
    pub preferred_family: AddressFamily,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum AddressDecisionKind {
    Start,
    Wait,
    Select,
    Timeout,
    Exhausted,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct AddressPlanDecision {
    pub kind: AddressDecisionKind,
    pub candidate_id: String,
    pub cancel: Vec<String>,
    pub wake_at_ms: Option<u64>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct AddressPlanResult {
    pub ordered_candidates: Vec<String>,
    pub decision: AddressPlanDecision,
}

fn decision(
    kind: AddressDecisionKind,
    candidate_id: impl Into<String>,
    cancel: Vec<String>,
    wake_at_ms: Option<u64>,
) -> AddressPlanDecision {
    AddressPlanDecision {
        kind,
        candidate_id: candidate_id.into(),
        cancel,
        wake_at_ms,
    }
}

fn ordered_candidates<'a>(
    candidates: &'a [AddressCandidate],
    attempts: &[AddressAttempt],
    context: &AddressPlanContext,
) -> Result<Vec<&'a AddressCandidate>, String> {
    let mut candidate_ids = HashSet::new();
    for candidate in candidates {
        if candidate.id.is_empty() || candidate.address.is_empty() {
            return Err("candidate id and address must not be empty".into());
        }
        if !candidate_ids.insert(candidate.id.as_str()) {
            return Err("candidate ids must be unique".into());
        }
    }

    let mut attempt_ids = HashSet::new();
    for attempt in attempts {
        if !attempt_ids.insert(attempt.candidate_id.as_str()) {
            return Err("attempt candidate ids must be unique".into());
        }
        if !candidate_ids.contains(attempt.candidate_id.as_str()) {
            return Err("attempt references an unknown candidate".into());
        }
        if attempt.state == AddressAttemptState::Succeeded && attempt.completed_at_ms.is_none() {
            return Err("successful attempt requires completed_at_ms".into());
        }
    }

    let mut ipv4 = Vec::new();
    let mut ipv6 = Vec::new();
    let mut seen_addresses = HashSet::new();
    for candidate in candidates {
        let already_started = attempt_ids.contains(candidate.id.as_str());
        let unexpired = candidate
            .expires_at_ms
            .map(|expiry| expiry > context.now_ms)
            .unwrap_or(true);
        if !already_started && !unexpired {
            continue;
        }
        let address_key = (candidate.family, candidate.address.as_str());
        if !seen_addresses.insert(address_key) {
            if already_started {
                return Err("attempt references a duplicate address candidate".into());
            }
            continue;
        }
        match candidate.family {
            AddressFamily::Ipv4 => ipv4.push(candidate),
            AddressFamily::Ipv6 => ipv6.push(candidate),
        }
    }

    let (preferred, alternate) = match context.preferred_family {
        AddressFamily::Ipv4 => (&ipv4, &ipv6),
        AddressFamily::Ipv6 => (&ipv6, &ipv4),
    };
    let mut ordered = Vec::with_capacity(preferred.len() + alternate.len());
    for index in 0..preferred.len().max(alternate.len()) {
        if let Some(candidate) = preferred.get(index) {
            ordered.push(*candidate);
        }
        if let Some(candidate) = alternate.get(index) {
            ordered.push(*candidate);
        }
    }
    Ok(ordered)
}

pub fn plan_address_attempts(
    candidates: &[AddressCandidate],
    attempts: &[AddressAttempt],
    context: &AddressPlanContext,
) -> Result<AddressPlanResult, String> {
    if context.max_concurrent < 1 {
        return Err("max_concurrent must be positive".into());
    }
    if context.deadline_at_ms <= context.started_at_ms {
        return Err("deadline must be after the race start".into());
    }

    let ordered = ordered_candidates(candidates, attempts, context)?;
    let ordered_ids: Vec<String> = ordered
        .iter()
        .map(|candidate| candidate.id.clone())
        .collect();
    let order_index: HashMap<&str, usize> = ordered_ids
        .iter()
        .enumerate()
        .map(|(index, id)| (id.as_str(), index))
        .collect();
    let attempts_by_id: HashMap<&str, &AddressAttempt> = attempts
        .iter()
        .map(|attempt| (attempt.candidate_id.as_str(), attempt))
        .collect();

    if attempts_by_id
        .keys()
        .any(|candidate_id| !order_index.contains_key(candidate_id))
    {
        return Err("attempt references a candidate unavailable to this plan".into());
    }

    let winner = attempts
        .iter()
        .filter(|attempt| attempt.state == AddressAttemptState::Succeeded)
        .min_by_key(|attempt| {
            (
                attempt.completed_at_ms.expect("validated completion"),
                order_index[attempt.candidate_id.as_str()],
            )
        });
    if let Some(winner) = winner {
        let cancel = ordered_ids
            .iter()
            .filter(|candidate_id| candidate_id.as_str() != winner.candidate_id)
            .filter(|candidate_id| {
                attempts_by_id
                    .get(candidate_id.as_str())
                    .map(|attempt| attempt.state == AddressAttemptState::Running)
                    .unwrap_or(false)
            })
            .cloned()
            .collect();
        return Ok(AddressPlanResult {
            ordered_candidates: ordered_ids,
            decision: decision(
                AddressDecisionKind::Select,
                winner.candidate_id.clone(),
                cancel,
                None,
            ),
        });
    }

    let running: Vec<&AddressAttempt> = attempts
        .iter()
        .filter(|attempt| attempt.state == AddressAttemptState::Running)
        .collect();
    if context.now_ms >= context.deadline_at_ms {
        let cancel = ordered_ids
            .iter()
            .filter(|candidate_id| {
                attempts_by_id
                    .get(candidate_id.as_str())
                    .map(|attempt| attempt.state == AddressAttemptState::Running)
                    .unwrap_or(false)
            })
            .cloned()
            .collect();
        return Ok(AddressPlanResult {
            ordered_candidates: ordered_ids,
            decision: decision(AddressDecisionKind::Timeout, "", cancel, None),
        });
    }

    let pending: Vec<String> = ordered_ids
        .iter()
        .filter(|candidate_id| !attempts_by_id.contains_key(candidate_id.as_str()))
        .cloned()
        .collect();
    if pending.is_empty() {
        let next = if running.is_empty() {
            decision(AddressDecisionKind::Exhausted, "", Vec::new(), None)
        } else {
            decision(
                AddressDecisionKind::Wait,
                "",
                Vec::new(),
                Some(context.deadline_at_ms),
            )
        };
        return Ok(AddressPlanResult {
            ordered_candidates: ordered_ids,
            decision: next,
        });
    }

    if running.is_empty() {
        return Ok(AddressPlanResult {
            ordered_candidates: ordered_ids,
            decision: decision(
                AddressDecisionKind::Start,
                pending[0].clone(),
                Vec::new(),
                None,
            ),
        });
    }

    if running.len() >= context.max_concurrent {
        return Ok(AddressPlanResult {
            ordered_candidates: ordered_ids,
            decision: decision(
                AddressDecisionKind::Wait,
                "",
                Vec::new(),
                Some(context.deadline_at_ms),
            ),
        });
    }

    let latest_start = attempts
        .iter()
        .map(|attempt| attempt.started_at_ms)
        .max()
        .unwrap_or(context.started_at_ms);
    let next_start_at = latest_start.saturating_add(context.stagger_ms);
    if context.now_ms >= next_start_at {
        return Ok(AddressPlanResult {
            ordered_candidates: ordered_ids,
            decision: decision(
                AddressDecisionKind::Start,
                pending[0].clone(),
                Vec::new(),
                None,
            ),
        });
    }

    Ok(AddressPlanResult {
        ordered_candidates: ordered_ids,
        decision: decision(
            AddressDecisionKind::Wait,
            "",
            Vec::new(),
            Some(next_start_at.min(context.deadline_at_ms)),
        ),
    })
}

File: spike/address_attempts.py
"""Pure deterministic address-attempt planning with no DNS or socket I/O."""

from dataclasses import dataclass
from typing import Optional


FAMILY_IPV4 = "ipv4"
FAMILY_IPV6 = "ipv6"

ATTEMPT_RUNNING = "running"
ATTEMPT_FAILED = "failed"
ATTEMPT_SUCCEEDED = "succeeded"
ATTEMPT_CANCELLED = "cancelled"

DECISION_START = "start"
DECISION_WAIT = "wait"
DECISION_SELECT = "select"
DECISION_TIMEOUT = "timeout"
DECISION_EXHAUSTED = "exhausted"


@dataclass(frozen=True)
class AddressCandidate:
    id: str
    family: str
    address: str
    source: str
    expires_at_ms: Optional[int] = None


@dataclass(frozen=True)
class AddressAttempt:
    candidate_id: str
    state: str
    started_at_ms: int
    completed_at_ms: Optional[int] = None


@dataclass(frozen=True)
class AddressPlanContext:
    now_ms: int
    started_at_ms: int
    deadline_at_ms: int
    stagger_ms: int
    max_concurrent: int
    preferred_family: str


@dataclass(frozen=True)
class AddressPlanDecision:
    kind: str
    candidate_id: str = ""
    cancel: tuple = ()
    wake_at_ms: Optional[int] = None


@dataclass(frozen=True)
class AddressPlanResult:
    ordered_candidates: tuple
    decision: AddressPlanDecision


def _validate_context(context):
    if context.preferred_family not in (FAMILY_IPV4, FAMILY_IPV6):
        raise ValueError("preferred_family must be ipv4 or ipv6")
    if context.max_concurrent < 1:
        raise ValueError("max_concurrent must be positive")
    if context.stagger_ms < 0:
        raise ValueError("stagger_ms must not be negative")
    if context.deadline_at_ms <= context.started_at_ms:
        raise ValueError("deadline must be after the race start")


def _ordered_candidates(candidates, attempts, context):
    candidate_ids = set()
    for candidate in candidates:
        if not candidate.id or not candidate.address:
            raise ValueError("candidate id and address must not be empty")
        if candidate.id in candidate_ids:
            raise ValueError("candidate ids must be unique")
        if candidate.family not in (FAMILY_IPV4, FAMILY_IPV6):
            raise ValueError("candidate family must be ipv4 or ipv6")
        candidate_ids.add(candidate.id)

    attempt_ids = set()
    for attempt in attempts:
        if attempt.candidate_id in attempt_ids:
            raise ValueError("attempt candidate ids must be unique")
        if attempt.candidate_id not in candidate_ids:
            raise ValueError("attempt references an unknown candidate")
        if attempt.state not in (
            ATTEMPT_RUNNING,
            ATTEMPT_FAILED,
            ATTEMPT_SUCCEEDED,
            ATTEMPT_CANCELLED,
        ):
            raise ValueError("unknown attempt state")
        if attempt.state == ATTEMPT_SUCCEEDED and attempt.completed_at_ms is None:
            raise ValueError("successful attempt requires completed_at_ms")
        attempt_ids.add(attempt.candidate_id)

    by_family = {FAMILY_IPV4: [], FAMILY_IPV6: []}
    seen_addresses = set()
    for candidate in candidates:
        already_started = candidate.id in attempt_ids
        unexpired = (
            candidate.expires_at_ms is None
            or candidate.expires_at_ms > context.now_ms
        )
        if not already_started and not unexpired:
            continue
        address_key = (candidate.family, candidate.address)
        if address_key in seen_addresses:
            if already_started:
                raise ValueError("attempt references a duplicate address candidate")
            continue
        seen_addresses.add(address_key)
        by_family[candidate.family].append(candidate)

    preferred = by_family[context.preferred_family]
    alternate_family = (
        FAMILY_IPV4 if context.preferred_family == FAMILY_IPV6 else FAMILY_IPV6
    )
    alternate = by_family[alternate_family]
    ordered = []
    for index in range(max(len(preferred), len(alternate))):
        if index < len(preferred):
            ordered.append(preferred[index])
        if index < len(alternate):
            ordered.append(alternate[index])
    return ordered


def plan_address_attempts(candidates, attempts, context):
    """Return the next bounded Happy-Eyeballs-style action for one snapshot."""
    _validate_context(context)
    ordered = _ordered_candidates(candidates, attempts, context)
    ordered_ids = tuple(candidate.id for candidate in ordered)
    order_index = {candidate_id: index for index, candidate_id in enumerate(ordered_ids)}
    attempts_by_id = {attempt.candidate_id: attempt for attempt in attempts}

    missing = set(attempts_by_id) - set(ordered_ids)
    if missing:
        raise ValueError("attempt references a candidate unavailable to this plan")

    successful = [
        attempt for attempt in attempts if attempt.state == ATTEMPT_SUCCEEDED
    ]
    if successful:
        winner = min(
            successful,
            key=lambda attempt: (
                attempt.completed_at_ms,
                order_index[attempt.candidate_id],
            ),
        )
        cancel = tuple(
            candidate_id
            for candidate_id in ordered_ids
            if candidate_id != winner.candidate_id
            and attempts_by_id.get(candidate_id)
            and attempts_by_id[candidate_id].state == ATTEMPT_RUNNING
        )
        return AddressPlanResult(
            ordered_ids,
            AddressPlanDecision(DECISION_SELECT, winner.candidate_id, cancel),
        )

    running = [attempt for attempt in attempts if attempt.state == ATTEMPT_RUNNING]
    if context.now_ms >= context.deadline_at_ms:
        cancel = tuple(
            candidate_id
            for candidate_id in ordered_ids
            if attempts_by_id.get(candidate_id)
            and attempts_by_id[candidate_id].state == ATTEMPT_RUNNING
        )
        return AddressPlanResult(
            ordered_ids,
            AddressPlanDecision(DECISION_TIMEOUT, cancel=cancel),
        )

    pending = [
        candidate_id for candidate_id in ordered_ids if candidate_id not in attempts_by_id
    ]
    if not pending:
        if running:
            return AddressPlanResult(
                ordered_ids,
                AddressPlanDecision(
                    DECISION_WAIT, wake_at_ms=context.deadline_at_ms
                ),
            )
        return AddressPlanResult(
            ordered_ids,
            AddressPlanDecision(DECISION_EXHAUSTED),
        )

    if not running:
        return AddressPlanResult(
            ordered_ids,
            AddressPlanDecision(DECISION_START, pending[0]),
        )

    if len(running) >= context.max_concurrent:
        return AddressPlanResult(
            ordered_ids,
            AddressPlanDecision(DECISION_WAIT, wake_at_ms=context.deadline_at_ms),
        )

    next_start_at = max(attempt.started_at_ms for attempt in attempts) + context.stagger_ms
    if context.now_ms >= next_start_at:
        return AddressPlanResult(
            ordered_ids,
            AddressPlanDecision(DECISION_START, pending[0]),
        )

    return AddressPlanResult(
        ordered_ids,
        AddressPlanDecision(
            DECISION_WAIT,
            wake_at_ms=min(next_start_at, context.deadline_at_ms),
        ),
    )
